text/0000-remove-shell-processes.md

```diff
@@ -0,0 +1,401 @@
+
+### Example 3 - A Script Process
+The follow custom script command:
+```
+docker run --entrypoint launcher <image> 'for opt in $JAVA_OPTS; do echo $opt; done'
+```
+will become the following, using the new platform API
+```
+docker run --entrypoint launcher <image> bash -c 'for opt in $JAVA_OPTS; do echo $opt; done'
+```
+
+## What About Profile Scripts?
+
+### `<layer>/profile.d/*`
+When a buildpack upgrades to the new buildpack API, buildpack-provided profile scripts will no longer be supported. Instead buildpacks can use `exec.d`. Most existing profiles scripts can be easily converted.
+
+**Example 1: `jkutner/sshd` **
+
+The following profile script example was taken from https://github.com/jkutner/sshd-buildpack/blob/master/sbin/sshd.sh
+```
+#!/usr/bin/env bash
+
+if [[ "${SSH_DISABLED:-}" != "true" ]]; then
+  ssh_layer=$(realpath $(dirname ${BASH_SOURCE[0]})/..)
+  ssh_dir=$(realpath $HOME/.ssh)
+
+  mkdir -p $ssh_dir
+  cat $ssh_layer/id_rsa.pub >> $ssh_dir/authorized_keys
+
+  cat << EOF >> $ssh_dir/sshd_config
+HostKey $ssh_layer/id_rsa
+AuthorizedKeysFile $ssh_dir/authorized_keys
+EOF
+
+  chmod 600 /workspace/.ssh/*
+
+  ssh_port=${SSH_PORT:-"2222"}
+  echo "at=sshd state=starting user=$(whoami) port=${ssh_port}"
+  /usr/sbin/sshd -f $ssh_dir/sshd_config -o "Port ${ssh_port}"
+fi
+```
+This script does not require any changes and would work as an `exec.d` or `profile.d`.
+
+**Example 2: paketo-buildpacks/node-engine **
+
+The following profile script example was taken from https://github.com/paketo-buildpacks/node-engine/blob/main/environment.go
+```
+if [[ -z "$MEMORY_AVAILABLE" ]]; then
+		memory_in_bytes="$(cat /sys/fs/cgroup/memory/memory.limit_in_bytes)"
+		MEMORY_AVAILABLE="$(( $memory_in_bytes / ( 1024 * 1024 ) ))"
+fi
+export MEMORY_AVAILABLE
+```
+
+The following revised version of this script would implement the `exec.d` interface
+```
+if [[ -z "$MEMORY_AVAILABLE" ]]; then
+		memory_in_bytes="$(cat /sys/fs/cgroup/memory/memory.limit_in_bytes)"
+		MEMORY_AVAILABLE="$(( $memory_in_bytes / ( 1024 * 1024 ) ))"
+fi
+
+cat << EOF >&3
+MEMORY_AVAILABLE = "${MEMORY_AVAILABLE}" 
+EOF
+```
+
+### `<app>/.profile`
+
+When a platform upgrades to the new platform API, user-provided profile scripts will no longer be supported. A buildpack could provide an identical or similar interface by detecting a `.profile` file in the app dir, wrapping it so that it implements the `exec.d` interface and add creating an executable with that layer. A platform that wishes to always provide that functionality could add a buildpack like this to every build. By making dynamic runtime environment modification a buildpack concern and not duplicating similar behavior in the platform API, we reduce the surface area and complexity of the platform API and thus improve normalization of the spec.
```

Comment:
Is this true even when the use `bash -c` in their command?

Reply:
Yes I believe this is still true. `exec.d` helpers currently run for both direct and shell processes.

Reply:
No, i mean if they use `bash -c` will `.profile` still be not be run? I feel like that might be unexpected.

Reply:
Yes, `.profile` will still not be run even when the command includes `bash -c`. Part of the goal here is to avoid doing anything complicated or magic. In order to run `.profile`s we would need to change the provided command from something like `['bash', '-c', 'echo hi']` to `['bash', '-c', 'source <app>/.profile && echo hi']`. Modifying the provided command if and only if it starts with `bash -c` feel like too much instrusive magic to me.